text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,401 @@
+
+Then the trait methods are rewritten so that they call the extern functions (using `as_ref()` to convert `&self` to a `&JsValue`),
+and they are marked as `#[inline]`.
+
+Now that we have defined the desired traits and methods, we can add the methods to a type like this:
+
+```rust
+#[wasm_bindgen]
+extern {
+    type EventTarget;
+}
+
+impl IEventTarget for EventTarget {}
+```
+
+```rust
+#[wasm_bindgen]
+extern {
+    type Node;
+}
+
+impl IEventTarget for Node {}
+impl INode for Node {}
+```
+
+Because the traits contain default implementations, this is all that is needed to make it work.
+
+## Mixins
+
+The above technique works great for class-based inheritance, but it can also be used to support mixins.
+
+As an example, the [`HTMLElement`](https://html.spec.whatwg.org/multipage/dom.html#htmlelement) class in
+WebIDL extends from `Element`, but *in addition* it includes various mixins, such as
+[`GlobalEventHandlers`](https://html.spec.whatwg.org/multipage/webappapis.html#globaleventhandlers) and
+[`HTMLOrSVGElement`](https://html.spec.whatwg.org/multipage/dom.html#htmlorsvgelement).
+
+This is essentially a form of multiple-inheritance, but thankfully Rust traits support multiple inheritance!
+
+```rust
+#[wasm_bindgen]
+pub trait IGlobalEventHandlers {
+    #[wasm_bindgen(setter)]
+    fn onabort(&self, listener: &Closure<FnMut(Event)>);
+
+    // And a bunch more methods...
+}
+
+#[wasm_bindgen]
+pub trait IHtmlOrSvgElement {
+    #[wasm_bindgen(getter)]
+    fn nonce(&self) -> JsString;
+
+    // And a bunch more methods...
+}
+
+#[wasm_bindgen]
+pub trait IHtmlElement: IElement + IGlobalEventHandlers + IHtmlOrSvgElement {
+    #[wasm_bindgen(getter = title)]
+    fn title(&self) -> JsString;
+
+    // And a bunch more methods...
+}
+```
+
+As you can see, the `IHtmlElement` trait inherits from `IElement`, `IGlobalEventHandlers`, and `IHtmlOrSvgElements`,
+thus achieving the multiple inheritance we need.
+
+And now we can implement the above traits on a type:
+
+```rust
+#[wasm_bindgen]
+extern {
+    type HtmlElement;
+}
+
+impl IGlobalEventHandlers for HtmlElement {}
+impl IHtmlOrSvgElement for HtmlElement {}
+impl IEventTarget for HtmlElement {}
+impl INode for HtmlElement {}
+impl IElement for HtmlElement {}
+impl IHtmlElement for HtmlElement {}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The `wasm_bindgen` attribute needs to be extended so it can be used on `trait`,
+this is a ***lot*** of extra complexity.
+
+In addition, because the methods are on traits, it is necessary for the Rust
+user to import the trait before they can use the methods:
+
+```rust
+let x: HtmlElement = ...;
+
+// Error, because the INode trait isn't imported
+x.append_child(y);
+```
+
+This drawback can be minimized by having a `traits` module which re-exports all
+of the traits. So that way the user can just put this at the top of their module:
+
+```rust
+// Now all of the methods work!
+use web_sys::traits::*;
```

Comment:
This would be a second `.rs` file emitted by the WebIDL frontend?

Reply:
It could also be an inline module, e.g.:
```rust
// ... all the bindings that we currently generate + trait definitions
pub mod traits {
    pub use super::{
        INode,
        IHtmlElement,
        // ...
    };
}
```
It would be easy enough to track each of the interfaces to add such an inline module at the end of the generated module.

Reply:
Regardless of whether it's a separate `.rs` or an inline module, either way I expect it to be auto-generated by the WebIDL generator.

Reply:
All I was saying is that it would be pretty straightforward to implement an extra traits module.